### ISSUE TITLE
DEP: Deprecate constructing dtypes from any object having a .dtype attribute

### ISSUE THREAD

**Issue by timhoffm** [numpy/numpy#25306]
### Proposed new feature or change:

I suggest to deprecate this feature. This is essentially already alluded in "we could consider deprecating even more" in https://github.com/numpy/numpy/pull/13578#issuecomment-755400474.

It seems too permissive and complicates code (e.g. #13003). Users should instead dereference the `.dtype` attribute explicitly ("explicit is better than implicit"), e.g.
``` 
np.array(data, dtype=obj.dtype)
# instead of
np.array(data, dtype=obj)
```

Also, the [current documentation](https://numpy.org/devdocs/reference/arrays.dtypes.html) is not exactly correct. It claims

> Any type object with a `.dtype` attribute

but this does not work (IMHO for good reasons) for ndarray objects.

**Comment by jakevdp** [comment]
Hi! I appreciate the motivation for this, but one data point: as proposed, this will make jax scalar type aliases (e.g. `jax.numpy.float32`, `jax.numpy.int32`, etc.) unusable as dtypes.

Some background: JAX made the choice early to not have distinct scalar type classes, but rather to represent scalars as 0-dimensional arrays. What this means is that when it comes to the `jax.numpy` API, we want `jnp.float32(0)` to be a function that returns a zero-dimensional `jax.Array`.

Unfortunately, a common pattern in numpy is to use scalar constructors as dtype arguments; e.g. `jnp.arange(4, dtype=jnp.float32)`. In order to support this, we've made the scalar constructors objects that have a `dtype` attribute, so that `np.dtype(jnp.float32)` will return `dtype('float32')`. If that mechanism is taken away, I suspect it would break a lot of existing JAX code.

Perhaps it would be possible to replace this with some other better-defined mechanism, e.g. using a `__dtype__` attribute to tell NumPy how `np.dtype` should treat a particular object?

**Comment by rkern** [comment]
Adding a designed-for-purpose mechanism like `.__dtype__`, _then_ deprecating the it-seemed-like-a-good-idea-at-the-time mechanism of inspecting the multipurpose `.dtype` makes sense to me.

**Comment by timhoffm** [comment]
Thanks for the comments. Makes sense. `__dtype__` may even be something to be formalized in the array API standard. Do you have any opinion on whether one should first bring this up there, or whether introducing it into numpy directly is more appropriate.

**Comment by seberg** [comment]
> __dtype__ may even be something to be formalized in the array API standard

I don't really see it fitting there.  We are talking about mixing different implementations and I would say it is better if it is `__numpy_dtype__` here anyway.

But yes, I agree with the idea that it is odd, but also agree that it needs a replacement since JAX and probably also pandas rely on it.

**Comment by jakevdp** [comment]
Circling back here: has there been any more discussion of replacing `.dtype` with `.__dtype__` to specify objects that are coercible to numpy dtype?

**Comment by seberg** [comment]
I don't think it matters, as we should just deprecate `.dtype` attribute lookup slowly when you don't find `__numpy_dtype__`, but would probably press the merge button on a PR adding it (with or without deprecation).
It seems like a pretty clear idea to try to be specific if we already need such logic...

**Comment by jakevdp** [comment]
> Adding a designed-for-purpose mechanism like `.__dtype__`, _then_ deprecating the it-seemed-like-a-good-idea-at-the-time mechanism of inspecting the multipurpose `.dtype` makes sense to me.

Could we revisit this idea of supporting a `__dtype__` protocol within `np.dtype`? Would it be better to bring this up on the mailing list?

**Comment by jorenham** [comment]
> > Adding a designed-for-purpose mechanism like `.__dtype__`, _then_ deprecating the it-seemed-like-a-good-idea-at-the-time mechanism of inspecting the multipurpose `.dtype` makes sense to me.
> 
> Could we revisit this idea of supporting a `__dtype__` protocol within `np.dtype`? Would it be better to bring this up on the mailing list?

Sounds like it would be a better fit for the Array API spec

**Comment by jakevdp** [comment]
> Sounds like it would be a better fit for the Array API spec

The array API spec tends to follow rather than lead – we rarely implement new features/APIs there, rather we formalize specifications for features that array libraries already support. In the case of dtype coercion, I think it makes the most sense to discuss this in NumPy first.

**Comment by seberg** [comment]
+1, we should add a new dunder.  I don't think we can actually deprecate `.dtype` lookup, at least for a very long time.

**Comment by jorenham** [comment]
I'm not sure if I like this. It sounds very ["xkcd 927"](https://xkcd.com/927/) to me:

<img width="500" height="283" alt="Image" src="https://github.com/user-attachments/assets/5104fb47-c200-4f9f-ab1c-f497c77de1ff" />

---

Currently the `numpy.typing.DTypeLike` type alias is already pretty massive. In the simplest of examples, this can already lead to sphaghetti error reports:

```pyi
from typing import Any
import numpy.typing as npt


def scalar(dtype: npt.DTypeLike) -> Any: ...

scalar(1)  # an `int` is not a `DTypeLike` (but a `type[int]` is)
```

Pulling this through mypy, and it'll report

```
demo.pyi:7: error: Argument 1 to "scalar" has incompatible type "int"; expected "type[Any] | dtype[Any] | _SupportsDType[dtype[Any]] | tuple[Any, Any] | list[Any] | _DTypeDict | str | None"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

, Pyright reports

```
/home/joren/Workspace/np-typing-stuff/demo.pyi
  /home/joren/Workspace/np-typing-stuff/demo.pyi:7:8 - error: Argument of type "Literal[1]" cannot be assigned to parameter "dtype" of type "DTypeLike" in function "scalar"
    Type "Literal[1]" is not assignable to type "DTypeLike"
      "Literal[1]" is not assignable to "type[Any]"
      "Literal[1]" is not assignable to "dtype[Any]"
      "Literal[1]" is incompatible with protocol "_SupportsDType[dtype[Any]]"
        "dtype" is not present
      "Literal[1]" is not assignable to "tuple[_DTypeLikeNested, _DTypeLikeNested]"
      "Literal[1]" is not assignable to "list[Any]"
      "Literal[1]" is not assignable to "_DTypeDict"
    ... (reportArgumentType)
1 error, 0 warnings, 0 informations
```

, and Pyrefly reports

```
ERROR Argument `Literal[1]` is not assignable to parameter `dtype` with type `_SupportsDType[dtype[Any]] | dtype[Any] | list[Any] | str | TypedDict[_DTypeDict] | tuple[Any, Any] | type[Any] | None` in function `scalar` [bad-argument-type]
 --> demo.pyi:7:8
  |
7 | scalar(1)
  |        ^
  |
  Protocol `_SupportsDType` requires attribute `dtype`
 INFO 1 error
```

Adding another another `typing.Protocol` for a `__dtype__: T` attribute to this will make this make this even more complicated.

---

And to be honest, `__dtype__` sounds like it's a solution in search of a problem. I'm sure there are *some* cases where it might help a bit. 

So given all that, I just don't think that the benefits outweigh the costs.

**Comment by seberg** [comment]
The point is that there are some libraries that have their own `.dtype` (e.g. torch, pandas, etc.) that are incompatible.  Only if we have a protocol (i.e. `__numpy_dtype__`) we can support converting to NumPy for that.

So yeah I think my point is that we have 0 protocols right now, because `.dtype` isn't a reasonable protocol to get to NumPy.

**Comment by jorenham** [comment]
> The point is that there are some libraries that have their own `.dtype` (e.g. torch, pandas, etc.) that are incompatible. Only if we have a protocol (i.e. `__numpy_dtype__`) we can support converting to NumPy for that.

`.__array__().dtype`?

**Comment by timhoffm** [comment]
In my understanding `__array__()` can be quite costly. Is it acceptable to have access to this information behind a costly function?

**Comment by seberg** [comment]
Yeah, it would be costly, but I think even more importantly... It may make sense to want to convert an XY dtype to a NumPy dtype, and `dtype.__array__()` makes no sense at all.

**Comment by jorenham** [comment]
If we know what the python type (e.g. bool, int, float, comnplex) is, the byteorder, and the itemsize, then we can figure out what numpy dtype it corresponds to. And the most obvious way place to define that would be in the array api dtype spec ([src](https://data-apis.org/array-api/latest/API_specification/data_types.html)), which currently is very under-specified.

**Comment by jorenham** [comment]
> The array API spec tends to follow rather than lead – we rarely implement new features/APIs there, rather we formalize specifications for features that array libraries already support. In the case of dtype coercion, I think it makes the most sense to discuss this in NumPy first.

The proposed `__dtype__` (that always returns a **numpy** dtype instance) cannot be adopted by the array-api, because it's specific to numpy. If they were to adopt it, it would return a dtype instance that follows the array api spec ([docs](https://data-apis.org/array-api/latest/API_specification/data_types.html)). And we just can't expect the array-api spec to add numpy as a required dependency.
So if we want this to be adopted by the array api, then I don't see how numpy would be able to "take the lead", as it will require changing the dtype spec of the array api.

**Comment by seberg** [comment]
@jorenham I am not sure where you getting at?

The point is that we want a way to convert a pandas dtype to a NumPy dtype (using pandas, I am not sure it applies, but I know others where it does).  Right now this works because `pandas_dtype.dtype` is a NumPy one but that is arguably strange?! Why should `dtype.dtype` be something completely different?

So right now, NumPy assumes that any `.dtype` is only reasonably a NumPy dtype, that is bad on two fronts:
1. `array.dtype` works, although it shouldn't be a valid to pass `dtype=array`.  (at least I think it does).
2. It forces the world to rotate around NumPy.  (I mean, it does, but we don't want to encode that ;)).

Switching over to `__numpy_dtype__` removes both ambiguities.

Now Array API?! Well, first of all, right now the array API has no concept of a what a dtype is, all it has is "there are names in the namespace".
Now if you would have a `__everyone_agrees_on_this_dtype__` you _might_ be able to argue to do only that and that is enough... But actually the best answer to that right now seems to have been "yeah, numpy is the best we got right now".
So yeah, if there was an obvious solution there that also works always (e.g. for structured dtypes) then sure, but I don't think that will exist.
Plus, I would argue that even _if_ we have an obvious solution, a `__numpy_dtype__` is still not a big deal to just have.
(But again, the main point was to not assume that `.dtype` is something we can rely on!)

---

> Would it be better to bring this up on the mailing list?

@jakevdp always good to ping the list, although I would phrase it as "we are planning to do this unless someone complains". I don't see a problem with it, I think it just (slowly) cleans things up to not assume the world rotates around NumPy (even if it does ;)).

**Comment by mattip** [comment]
Does this make sense?
- Encourage users to specify `np.array(data, dtype=obj.__numpy_dtype__)` in the documentation and in downstream libraries with the caveat that it might take time for this to propogate out.
- In `np.array(data, dtype=obj)` take the first successful of
  1 . Error if `obj` is a `ndarray` or `ndarray` subtype
  2 . Look for `isinstance(obj, np.dtype)`
  3 . Look for `obj.__numpy_dtype__` (and error if `not isinstance(obj.__numpy_dtype__, np.dtype)`)
  5 . Look for `obj.dtype` like today
  6 . Convert a Python `int`, `float` or `bool` to the corresponding NumPy dtype
  7 . Use `dtype=object`

We could then slowly deprecate (5) over time. If the ArrayAPI adopts a more generic `__dtype__` in some future version, we could add a "(4) look for `obj.__dtype__` and convert it according to the ArrayAPI spec"

**Comment by seberg** [comment]
Yes, agreed, that would be my view.  I _think_ it seems we already do 1 actually but this won't kick in for other array-like classes.  Maybe we could check for `__array__` or so but not sure we should.

But yeah, I would want to add checking for `__numpy_dtype__` before checkingi for a `dtype` attribute.  Then in some future we can deprecate the `dtype` attribute lookup.

**Comment by timhoffm** [comment]
> Encourage users to specify `np.array(data, dtype=obj.__numpy_dtype__)`

I think we rather want to encourage `np.array(data, dtype=obj.dtype)`. The `dtype` parameter should accept any *dtype-like* object. What *dtype-like* means is encoded in the enumeration [above](https://github.com/numpy/numpy/issues/25306#issuecomment-3501275259) (enventually with 5 being removed).

For example, currently you can do `series = pd.Series([1, 2]); np.array(data, dtype=series)`. That should be an error in the far future, and users should switch to `np.array(data, dtype=series.dtype)`, which will work right now already because `series.dtype` is already a numpy dtype.

Likewise, for `t = torch.tensor([1, 2])` we'd want `np.array(data, dtype=t.dtype)`. Since `t.dtype` is not a numpy dtype the torch dtype must somehow communicate what the corresponding numpy dtype is, and that what `__numpy_dtype__` will be introduced for.

In summary, array-like objects (series, tensor, etc.) should not have `__numpy_dtype__`. They may specify `obj.dtype` which can return any kind of library-specific dtype. These library-specific dtypes may define `__numpy_dtype__`. E.g. `torch.int64.__numpy_dtype__ = np.int64`.

**Comment by jakevdp** [comment]
> Does this make sense?

Yes, I think this is mostly great – number 1-4 in sequence, followed by a deprecated 5

>   6 . Convert a Python `int`, `float` or `bool` to the corresponding NumPy dtype
>   7 . Use `dtype=object`

If I'm understanding you correctly, this is where I disagree. I think `np.array([1, 2, 3], dtype=SomeArbitraryObject)` should error rather than falling back to `dtype=object`.

**Comment by seberg** [comment]
I think matti meant `int`, `float` as the types, not instances.  In fact they are checked first I think and honestly, that is probably OK.
And yes, of course users should *never* write `__numpy_dtype__`.

I suppose we actually can deprecate `.dtype` a bit quicker since downstream can just have both for a while, it adds very little overall effort.

**Comment by seberg** [comment]
A lot more tedious then I thought, because we have a mess with a `try: numpy.dtype(obj.dtype)/except` having a meaning of "if this fails then ignore the attribute"...
But I opened gh-30179.  We could maybe make it a bit nicer by making `__numpy_dtype__ = None` not add any details beyond "can't convert to a dtype".

(I used a ValueError, because I think that is what we use generally, although TypeError would feel slightly more to the point.  PR needs a release note, but I think it is ready otherwise.)

**Comment by seberg** [comment]
The PR means that you can implement `__numpy_dtype__` as a downstream ArrayLike object to prevent conversion to a dtype via `.dtype` (and raise a custom error if you like).  I think that is probably good enough.  If you would want to softly change that, it would be downstream who has to add a deprecation warning to it's `__numpy_dtype__`.
